### PR TITLE
Optimized conclude

### PR DESCRIPTION
**Pull request by geoknee**
Test coverage:

`concludeFromOpen`: 
- [x] 'It accepts a valid concludeFromOpen tx (n states) and sets the channel storage correctly';
- [x] 'It accepts a valid concludeFromOpen tx (1 state) and sets the channel storage correctly';
- [x] 'It accepts a valid concludeFromOpen tx (1 state, cleared challenge exists) and sets the channel storage correctly';
- [x] 'It reverts a concludeFromOpen tx when the declaredTurnNumRecord = 0 and incorrect';
- [x] 'It reverts a concludeFromOpen tx when the declaredTurnNumRecord > 0 and incorrect';
- [x] 'It reverts a concludeFromOpen tx when there is an ongoing challenge';
- [x] 'It reverts a concludeFromOpen tx when the outcome is already finalized';

Consumes no more than 97K gas.

`concludeFromChallenge`:

- [x] 'It accepts a valid concludeFromChallenge tx (n states) and sets the channel storage correctly';
- [x] 'It reverts a concludeFromChallenge tx when there is no challenge ongoing (turnNumRecord = 0)';
- [x] 'It reverts a concludeFromChallenge tx when there is no challenge ongoing (challenge cleared)';
- [x] 'It reverts a concludeFromChallenge tx when the outcome is already finalized';

Consumes no more than 93K gas

### CODE REVIEW

**Comment by geoknee** [comment]
As is, only the `outcomeHash` is supplied to these methods. I wonder if we should consider passing in the hydrated `outcome` data, and then hashing it inside the function? That would allow hydrated outcome information to be emitted in the `Concluded` event, which will be needed in order to `pushOutcome`.

You might argue that all participants should already have this information, but they need only have signed the `stateHash`, which itself is the hash of `outcomeHash := keccak256(outcome)` and some other data. So they needn't known the outcome in order to approve it.

Alternatively, we just make clear that you should never sign any state unless you know the preimage of the `outcomeHash`.

File: contracts/OptimizedForceMove.sol [hunk 1]
```diff
@@ -409,6 +409,101 @@ contract OptimizedForceMove {
...
+        uint8 numStates,
+        uint8[] memory whoSignedWhat,
+        Signature[] memory sigs
+    ) public {
+        // Calculate channelId from fixed part
+        bytes32 channelId = keccak256(
+            abi.encode(fixedPart.chainId, fixedPart.participants, fixedPart.channelNonce)
+        );
+
+        if (turnNumRecord == 0) {
```
Comment (tomclose) [inline contracts/OptimizedForceMove.sol]: I noticed that this isn't the same as the equivalent check in `forceMove`, where you did:
```
if (channelStorageHashes[channelId] != bytes32(0)) {
  // ....
```
Was this deliberate?

**Comment by tomclose** [comment]
I think we can probably leave it as is. I think the time to add the full outcome would be if we want to add an optimized `concludeAndPush` method, when we create the Nitro Adjudicator.

I also think it should (hopefully) already be clear from the protocol that you shouldn't sign a state unless you have all the parts that comprise that state, which will include the outcome.

File: contracts/OptimizedForceMove.sol (hunk 1 above)
Comment (geoknee) [reply contracts/OptimizedForceMove.sol]: It was not deliberate; in edbd8d3  I have now used the `forceMove` version which is slightly better for gas consumption.

**Pull request merged**